#3: implement component rules C3-C7

vlad961 opened the pull request:
#3 
Implemented Component rules C3 and C4 with Archunit.

vlad961 committed:
Added first draft for C2:
Business architecture definition of dependencies test.

vlad961 committed:
Added first draft for C3:
Components service layers should not depend on each other.

vlad961 committed:
Added first C4 draft.
NOT READY!

vlad961 committed:
Added new draft for C4Layer2Logic4ComponentTest

NicolasVanBellen committed:
First commits of rule C5 - C7

NicolasVanBellen committed:
First commits for rules C5 - C7

NicolasVanBellen committed:
Merge branch 'add-component-rules' of https://github.com/NicolasVanBellen/archunit into add-component-rules

NicolasVanBellen committed:
almost final version of rules C5 - C7

vlad961 committed:
Refactored C4 Rule.

vlad961 committed:
Refactored C3:
Added a more concise error message.

vlad961 committed:
Refactored C2

vlad961 committed:
Refactoring C4 Rule + Delete C2 Rule

Deleted C2 rule since it is obsolete. It will be covered by #21.
Refactoring by automatic code-formatting of devonfw-ide.

vlad961 committed:
Merge branch 'devonfw-sample:master' into add-component-rules

vlad961 committed:
Update ComponentRuleC4LayerService2Logic4ComponentTest.java

Minor refactoring.
Changed ArchCondition to final.
Renamed boolean dependencyAllowed to isAllowedDependency.

NicolasVanBellen committed:
Merge branch 'vlad961:add-component-rules' into add-component-rules

NicolasVanBellen committed:
Output changes

vlad961 committed:
Merge pull request #2 from NicolasVanBellen/add-component-rules

Add component rules C5-C7

vlad961 committed:
Refactored C3 and C6

HOTFIX general component use.

Co-Authored-By: NicolasVanBellen <74297820+NicolasVanBellen@users.noreply.github.com>

vlad961 committed:
Merge branch 'devonfw-sample:master' into add-component-rules

vlad961 committed:
Merged all component rules into one file.

vlad961 committed:
Update ComponentRuleTest.java

@NicolasVanBellen
Adjusted component rules C3-C7 to use the newly introduced PackageStructure type.

Co-Authored-By: NicolasVanBellen <74297820+NicolasVanBellen@users.noreply.github.com>

hohwille commented on src/test/java/com/devonfw/sample/archunit/ComponentRuleTest.java:
```diff
@@ -0,0 +1,263 @@
+     */
+    @ArchTest
+    public static final ArchRule C3_no_service_layer_depends_on_service_layer_of_another_component = slices()
+            .matching("..archunit.(*).service..")
```
Same as PR #37

hohwille replied:
Can you simply remove `archunit.`?

vlad961 replied:
Removed.

hohwille commented on src/test/java/com/devonfw/sample/archunit/ComponentRuleTest.java:
```diff
@@ -0,0 +1,263 @@
+    public static final ArchRule C3_no_service_layer_depends_on_service_layer_of_another_component = slices()
+            .matching("..archunit.(*).service..")
+            .namingSlices("$1 service").should().notDependOnEachOther()
+            .ignoreDependency(alwaysTrue(), nameMatching(".*general.*"))
```
This will also match to e.g. `com.devonfw.sample.archunit.component.service.foogeneral`.
I guess applying `PackageStructure` to all these places will solve lots of problems.

vlad961 replied:
~~Adjusted the ignoreDependency like your [proposal for] (https://github.com/devonfw-sample/archunit/pull/37#discussion_r1135177629)~~

~~`.matching("..(*).service..").namingSlices("$1 service").should().notDependOnEachOther()
      .ignoreDependency(alwaysTrue(), nameMatching(".general.(common|service|logic|dataaccess|batch|client).."))`~~

Reimplemented rules C3 and C6 due to the erroneous use of ignoreDependency().
A custom ArchCondition combined with `PackageStructure `is utilized now to enfore rules C3 and C6.

hohwille commented on src/test/java/com/devonfw/sample/archunit/ComponentRuleTest.java:
```diff
@@ -0,0 +1,263 @@
+        public void check(JavaClass sourceClass, ConditionEvents events) {
+
+            String sourceClassPackageName = sourceClass.getPackageName();
+            PackageStructure sourcePkg = PackageStructure.of(sourceClassPackageName);
```
FYI: You have even created a method `PackageStructure.of(JavaClass)` that seems perfect here.

vlad961 replied:
Utilizing the JavaClass method for all rules now.

hohwille commented on src/test/java/com/devonfw/sample/archunit/ComponentRuleTest.java:
```diff
@@ -0,0 +1,263 @@
+            // Check all project components' service layers except the default component for
+            // noncompliant dependencies towards
+            // other components' logic layers.
+            if (sourcePkg.isLayerService() && !sourceClassComponent.equals("")
```
FYI: I do not see any case how `isLayerService()` could be `true` while  `sourceClassComponent.equals("")` as Java does not allow empty package segments and OS also does not allow folders with empty names.

vlad961 replied:
Deleted the `sourceClassComponent.equals("")` conditions throughout all rules.

hohwille commented on src/test/java/com/devonfw/sample/archunit/ComponentRuleTest.java:
```diff
@@ -0,0 +1,263 @@
+            // noncompliant dependencies towards
+            // other components' logic layers.
+            if (sourcePkg.isLayerService() && !sourceClassComponent.equals("")
+                    && !sourceClassComponent.equals("general")) {
```
Ah, I missed this case in `PackageStructure` but as we have to check for `general` in multiple places there should be a constant and also a getter (`isComponentGeneral()`) for this to avoid redundancies.

vlad961 replied:
I added a constant `public final DEFAULT_COMPONENT = "general";`
and the `isComponenetGeneral()` getter inside **_PackageStructure.java_**

hohwille commented on src/test/java/com/devonfw/sample/archunit/ComponentRuleTest.java:
```diff
@@ -0,0 +1,263 @@
+     * @param targetClass Target JavaClass to check if dependencies from sourceClass
+     *                    towards it are allowed.
+     */
+    public static boolean isComponentServiceLayerCompliant(JavaClass sourceClass, JavaClass targetClass) {
```
FYI: method is only internal and should not be `public`.

vlad961 replied:
Changed all helper methods into `private`

hohwille commented on src/test/java/com/devonfw/sample/archunit/ComponentRuleTest.java:
```diff
@@ -0,0 +1,263 @@
+     */
+    public static boolean isComponentServiceLayerCompliant(JavaClass sourceClass, JavaClass targetClass) {
+
+        boolean isAllowed = false;
+        PackageStructure sourcePkg = PackageStructure.of(sourceClass.getPackageName());
+        PackageStructure targetPkg = PackageStructure.of(targetClass.getPackageName());
```
why passing the JavaClass objects to this method and parsing them to `PackageStructure`s here if you already have them parsed before?

vlad961 replied:
Readjusted all helper method signatures to use PackageStructure types.

hohwille commented on src/test/java/com/devonfw/sample/archunit/ComponentRuleTest.java:
```diff
@@ -0,0 +1,263 @@
+     * Dependency of a components logic layer towards the same components dataaccess
+     * or common layer is allowed.
+     */
+    public static boolean isComponentLogicLayerCompliant(JavaClass sourceClass, JavaClass targetClass) {
+        boolean isAllowed = false;
+        PackageStructure sourcePkg = PackageStructure.of(sourceClass.getPackageName());
+        PackageStructure targetPkg = PackageStructure.of(targetClass.getPackageName());
```
Same here:
```
private static boolean isComponentLogicLayerCompliant(PackageStructure sourcePkg, PackageStructure targetPkg) {
```

vlad961 replied:
Changed method to `private`.
Readjusted the method signature to use `PackageStructure` types.

hohwille commented on src/test/java/com/devonfw/sample/archunit/ComponentRuleTest.java:
```diff
@@ -0,0 +1,263 @@
+     * Dependency of a components batch layer towards the same components logic or
+     * common layer is allowed.
+     */
+    public static boolean isComponentBatchLayerCompliant(JavaClass sourceClass, JavaClass targetClass) {
+        boolean isAllowed = false;
+        PackageStructure sourcePkg = PackageStructure.of(sourceClass.getPackageName());
+        PackageStructure targetPkg = PackageStructure.of(targetClass.getPackageName());
```
and here:
```
private static boolean isComponentBatchLayerCompliant(PackageStructure sourcePkg, PackageStructure targetPkg) {
```

vlad961 replied:
Changed method to `private`.
Readjusted the method signature to use `PackageStructure` types.

vlad961 committed:
Merge branch 'devonfw-sample:master' into add-component-rules

vlad961 committed:
Reworked Component Rules

Updated PackageStructure.java:
- Added a projects architecture DEFAULT_COMPONENT field.
- Added a isComponentGeneral() getter.

Updated ComponentRuleTest.java:
- Reimplemented C3 and C6 using a custom ArchCondition + PackageStructure
- Resolved all review annotation of @hohwille

Co-Authored-By: NicolasVanBellen <74297820+NicolasVanBellen@users.noreply.github.com>

hohwille committed:
Merge branch 'master' into add-component-rules

hohwille commented on src/test/java/com/devonfw/sample/archunit/ComponentRuleTest.java:
```diff
@@ -0,0 +1,328 @@
+          if (targetPkg.isLayerService() && !isAllowedDependency) {
+            String message = String.format(
+                "'%s.%s' is dependend on '%s.%s'. Violated in: (%s). Dependency towards (%s)", sourceClassComponent,
+                sourceClassLayer, targetClassComponent, targetClassLayer, sourceClass.getDescription(),
```
FYI: Due to the surrounding if condition the layer is not dynamic here but has to be `service`. So nothing is wrong here but as long as you are not extracting and reusing this code for other rules it can be slightly confusing as you could also hard-code the layer here and so I start thinking "why is it filled in as a dynamic variable then? Do I miss something?".

However, it may make a lot of sense to share the same format of error message for all component rules. Therefore you could extract the code to create this message and add an according event to a private method shared by all rules.

vlad961 replied:
I made major changes via DescribedPredicates which filters the previously used sourcePkg layers.
Afterward, ArchConditions check for targetPkg violations.

I added a `private static String composeViolationMessage(JavaClass sourceClass, JavaClass targetClass, PackageStructure sourcePkg, PackageStructure targetPkg)` method shared by all rules to use the same format of error message.

hohwille commented on src/test/java/com/devonfw/sample/archunit/ComponentRuleTest.java:
```diff
@@ -0,0 +1,328 @@
+public class ComponentRuleTest {
+
+  public static final ArchCondition<JavaClass> haveComponentServiceLayerDependingOnDiffComponentsServiceLayer = new ArchCondition<JavaClass>(
+      "have dependencies, towards another components service layer (Rule-C3).") {
```
I was slightly confused by this description in combination with the implementation that does not match it.
So I looked back into the original C3 rule and found this violation message:
`Code from service layer shall not depend on service layer of a different component.`

This is a very different message. If you change the text to a different meaning this can cause a lot of confusion to the users of this rule.

OK. In Line 57 I now found exactly this message. Still I am confused by this message here. Also I am unsure what a message for a condition is used for (where would I see this if the rule fails?) and what is the difference between `as("message")` and `because("message")` in ArchUnit...

vlad961 replied:
Using `as("message")` enforces printing only the `message `string.
Because concatenates the fluent API messages.
E.g.:

```
noClasses()
      .that(DescribedPredicate("message1"))
      .should(ArchCondition("message2"))
      .because("message3")
```


would result in _"no classes that message1 should message2, because message3"_.

I reworked my implementation and utilize now DescribedPredicates and ArchConditions.
There are more appropriate messages inside DescribedPredicates and ArchConditions now (which help to understand the code better). Nevertheless, I use `as("message")` to return a shorter error message.

hohwille commented on src/test/java/com/devonfw/sample/archunit/ComponentRuleTest.java:
```diff
@@ -0,0 +1,328 @@
+
+      // Check all project components' service layers except the default component for noncompliant dependencies towards
+      // other components' service layers.
+      if (sourcePkg.isLayerService() && !sourcePkg.isComponentGeneral()) {
```
where does the `!sourcePkg.isComponentGeneral()` come from? I do not get this.

vlad961 replied:
I introduced a `public static final String DEFAULT_COMPONENT="general"` field inside our `PackageStructure` and a corresponding `PackageStructure.isComponentGeneral()` getter, representing the project's default component to simplify the component rules.

hohwille commented on src/test/java/com/devonfw/sample/archunit/ComponentRuleTest.java:
```diff
@@ -0,0 +1,328 @@
+      isAllowed = true;
+    }
+    // Components may always depend on the default business component.
+    if (targetPkg.isComponentGeneral() && (targetPkg.isLayerLogic() || targetPkg.isLayerCommon())) {
```
I start thinking that with `general` we have a misunderstanding.
There are two dimensions:
1. An architect defines which component may potentially depend on which other component. This is part of the "business architecture" and can not be defined in a generic way. In `sonar-devon4j-plugin` we allowed to define this in the configuration file `architecture.json`, because the rules had to be fixed and provided as a released plugin artifact where the code was a black box to the user. In our ArchUnit sample this is different: We provide an example as template that can be copy&pasted into any project and adapted to the project needs. In this way, we can provide a definition of the business architecture for our sample. This might be tricky so far as we only have one real component yet (task). There are however, the following aspects that devonfw pre-defines as a best-practice for the business architecture: There is or may be a component `general` and that component must NOT have any dependency on other components (in our example access from `general` to `task` is always forbidden) AND all other components are always allowed to depend on `general`. In other words, `general` is the fundamental cross-cutting component and also it is the analogy to what `common` is for the layers. Additional there once was another predefined component `app` that on other component may depend on and that may potentially depend on all other components. This was designed for composition and aggregation code. As most projects never made use of this concept, we stopped having this build into devonfw. But the final conclusion for 1. is: It only defines what is disallowed but not fully defines what is allowed.
2. If component A is allowed to access component B according to 1 with A!=B. then there are additional rules defining what layers in A may access which layers in B. Only with the combination of 1. and 2. we can finally define what is allowed and what is forbidden but 1. and 2. can be implemented as separate rules rather independently.

vlad961 replied:
I have added an ArchRule that checks whether the `general` component does not depend on any other component.

For 1: I have changed the tests to check for disallowed dependencies rather than what is allowed.

hohwille reviewed (commented):
Now, I understand - I did already add all the review comments but they did not become visible because I did not finish the review.... So here are my recent comments.

hohwille committed:
Merge branch 'master' into add-component-rules

vlad961 committed:
Update ComponentRuleTest.java

Reworked component rules according to recent review.
- added DescribedPredicates
- added composeViolationMessage method
- added test for general component

hohwille reviewed (approved):
@vlad961 Wow. You perfectly nailed it and even addressed all my nice to have comments and perfectly implemented my idea for improvement. The rules now are beautiful to read. 👍 

Ready to merge. When implementing #19 we can still avoid some more redundancy and make this nice work available to all other rules.
